src/layers/GeomodelLabelsLayer.ts

```diff
@@ -260,15 +333,45 @@ export class GeomodelLabelsLayer extends CanvasLayer {
     return `#${hexString}`;
   }
 
-  calcPos(data: number[][], offset: number, count: number, step: number, topLimit: number = null, bottomLimit: number = null): Vector2 {
+  calcPos(
+    data: number[][],
+    offset: number,
+    count: number,
+    step: number,
+    topLimit: number = null,
+    bottomLimit: number = null,
+    altData: number[][] = null,
+    surfaces: any[] = null,
+    index: number = null,
+  ): Vector2 {
     const pos = Vector2.zero.mutable;
     let samples = 0;
     for (let i = 0; i < count; i++) {
       const x = offset + i * step;
 
       const y = findSampleAtPos(data, x, topLimit, bottomLimit);
       if (y) {
-        pos.add(x, y);
+        let altY;
+        if (altData) {
+          // Find sample from passed in surface
+          altY = findSampleAtPos(altData, x, topLimit, bottomLimit);
+          if (altY == null && surfaces && index != null) {
+            //Find topmost surface after current which gives us data
+            let si = index + 1;
+            while (altY == null && si < surfaces.length) {
+              const altSurface = surfaces[si++];
+              altY = findSampleAtPos(
+                altSurface.data.map((d: any) => [d[0], d[1]]),
+                x,
+                topLimit,
+                bottomLimit,
+              );
+            }
+          }
+        }
```

Comment:
Quite a lot of nesting going on here. How about splitting up into small functions and try to reduce levels of nesting?

Reply:
I'll have a look at this

Reply:
Cleaned up